core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java

```diff
@@ -49,9 +50,25 @@
  * A <code>SqlDialect</code> implementation for the MySQL database.
  */
 public class MysqlSqlDialect extends SqlDialect {
+
+  /** MySQL type system. */
+  private static final RelDataTypeSystem MYSQL_TYPE_SYSTEM =
+      new RelDataTypeSystemImpl() {
+        @Override public int getMaxPrecision(SqlTypeName typeName) {
+          switch (typeName) {
+          case CHAR:
+          case VARCHAR:
```

Comment:
This seems to override the max precision for VARCHAR to same value.
I just checked the documentation and value seems to be higher than that:
https://dev.mysql.com/doc/refman/5.7/en/char.html
Can we change it accordingly (and update tests)?

Reply:
calcite's mysql dialect rewrites VARCHAR into CHAR (not yet sure why), as a result getMaxPrecision uses same max precision for both

Reply:
`@Override public SqlNode getCastSpec(RelDataType type) {
    switch (type.getSqlTypeName()) {
    case VARCHAR:
      // MySQL doesn't have a VARCHAR type, only CHAR.
      return new SqlDataTypeSpec(
          new SqlBasicTypeNameSpec(SqlTypeName.CHAR, type.getPrecision(), SqlParserPos.ZERO),
          SqlParserPos.ZERO);`

Reply:
Even if CAST to VARCHAR is not allowed as per the comment below, I believe the max precision in the type system should be set correctly, since other methods may rely on it.

Reply:
Addressed this in latest push